// smart-education-backend/controllers/userController.js
const User = require('../models/User');

// @desc    Get all users
// @route   GET /api/users
// @access  Private/Admin
const getUsers = async (req, res) => {
  try {
    const users = await User.find().select('-password').sort({ createdAt: -1 });
    res.json(users);
  } catch (error) {
    console.error('Get users error:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

// @desc    Get users by role
// @route   GET /api/users/role/:role
// @access  Private/Admin or Teacher (for students only)
const getUsersByRole = async (req, res) => {
  try {
    const { role } = req.params;
    
    if (!['admin', 'teacher', 'student'].includes(role)) {
      return res.status(400).json({ message: 'Invalid role' });
    }

    // Teachers can only view students
    if (req.user.role === 'teacher' && role !== 'student') {
      return res.status(403).json({ message: 'Teachers can only view students' });
    }
    
    const users = await User.find({ role }).select('-password').sort({ createdAt: -1 });
    res.json(users);
  } catch (error) {
    console.error('Get users by role error:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

// @desc    Create user (Admin or Teacher)
// @route   POST /api/users
// @access  Private/Admin (all roles) or Teacher (students only)
const createUser = async (req, res) => {
  try {
    const { name, email, password, role } = req.body;

    // Validate input
    if (!name || !email || !password || !role) {
      return res.status(400).json({ message: 'Please provide all required fields' });
    }

    // Validate role
    if (!['admin', 'teacher', 'student'].includes(role)) {
      return res.status(400).json({ message: 'Invalid role' });
    }

    // Teachers can only create students
    if (req.user.role === 'teacher' && role !== 'student') {
      return res.status(403).json({ message: 'Teachers can only create student accounts' });
    }

    // Check if user exists
    const userExists = await User.findOne({ email });
    if (userExists) {
      return res.status(400).json({ message: 'User with this email already exists' });
    }

    // Create user
    const user = await User.create({
      name,
      email,
      password,
      role
    });

    res.status(201).json({
      _id: user._id,
      name: user.name,
      email: user.email,
      role: user.role,
      createdAt: user.createdAt
    });
  } catch (error) {
    // Handle validation errors
    if (error.name === 'ValidationError') {
      const errors = Object.fromEntries(
        Object.entries(error.errors || {}).map(([key, val]) => [key, val.message])
      );
      return res.status(400).json({ message: 'Validation failed', errors });
    }

    console.error('Create user error:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

// @desc    Delete user
// @route   DELETE /api/users/:id
// @access  Private/Admin or Teacher (students only)
const deleteUser = async (req, res) => {
  try {
    const user = await User.findById(req.params.id);

    if (!user) {
      return res.status(404).json({ message: 'User not found' });
    }

    // Prevent deleting themselves
    if (user._id.toString() === req.user._id.toString()) {
      return res.status(400).json({ message: 'Cannot delete your own account' });
    }

    // Teachers can only delete students
    if (req.user.role === 'teacher' && user.role !== 'student') {
      return res.status(403).json({ message: 'Teachers can only delete student accounts' });
    }

    await User.findByIdAndDelete(req.params.id);
    res.json({ message: 'User deleted successfully' });
  } catch (error) {
    console.error('Delete user error:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

// @desc    Get dashboard stats
// @route   GET /api/users/stats
// @access  Private/Admin or Teacher
const getStats = async (req, res) => {
  try {
    const totalUsers = await User.countDocuments();
    const totalStudents = await User.countDocuments({ role: 'student' });
    const totalTeachers = await User.countDocuments({ role: 'teacher' });
    const totalAdmins = await User.countDocuments({ role: 'admin' });

    res.json({
      totalUsers,
      totalStudents,
      totalTeachers,
      totalAdmins
    });
  } catch (error) {
    console.error('Get stats error:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

// @desc    Get recent platform activity (latest users, courses, materials, doubts)
// @route   GET /api/users/activity
// @access  Private/Admin
const getRecentActivity = async (req, res) => {
  try {
    const Course = require('../models/Course');
    const Material = require('../models/Material');
    const Doubt = require('../models/Doubt');

    const [recentUsers, recentCourses, recentMaterials, recentDoubts] = await Promise.all([
      User.find().select('name role createdAt').sort({ createdAt: -1 }).limit(5),
      Course.find().select('title createdBy createdAt').populate('createdBy', 'name').sort({ createdAt: -1 }).limit(5),
      Material.find().select('title subject uploadedBy createdAt').populate('uploadedBy', 'name').sort({ createdAt: -1 }).limit(5),
      Doubt.find().select('question askedBy status createdAt').populate('askedBy', 'name').sort({ createdAt: -1 }).limit(5)
    ]);

    // Merge into a single timeline sorted by date
    const activities = [
      ...recentUsers.map(u => ({
        type: 'user',
        title: `${u.name} joined as ${u.role}`,
        detail: `New ${u.role} account created`,
        date: u.createdAt
      })),
      ...recentCourses.map(c => ({
        type: 'course',
        title: `Course "${c.title}" created`,
        detail: `By ${c.createdBy?.name || 'Unknown'}`,
        date: c.createdAt
      })),
      ...recentMaterials.map(m => ({
        type: 'material',
        title: `Material "${m.title}" uploaded`,
        detail: `${m.subject} • By ${m.uploadedBy?.name || 'Unknown'}`,
        date: m.createdAt
      })),
      ...recentDoubts.map(d => ({
        type: 'doubt',
        title: `Doubt asked by ${d.askedBy?.name || 'Student'}`,
        detail: d.question?.substring(0, 60) + (d.question?.length > 60 ? '...' : ''),
        date: d.createdAt
      }))
    ].sort((a, b) => new Date(b.date) - new Date(a.date)).slice(0, 10);

    res.json({ activities });
  } catch (error) {
    console.error('Get recent activity error:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

module.exports = { getUsers, getUsersByRole, createUser, deleteUser, getStats, getRecentActivity };
